### PR TITLE
feat: filter resource with role based access (AM-2)

### PR DESCRIPTION
**Pull request by swalker125**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/AM-2

### Change description ###

Updates to filterResource. On empty ExplicitAccess now checks role based access. If no access returns null. If role based access with explicit access type, returns null. If role based access with role based access type, returns permissions and filters resource json.

Added envelope v2 https://tools.hmcts.net/confluence/display/AM/005+Envelope+Structure to filterResource method.

### CODE REVIEW

File: am-lib/src/main/java/uk/gov/hmcts/reform/amlib/models/Resource.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+
+@Data
+@Builder
+@AllArgsConstructor
+public class Resource {
+    private final String resourceId;
+    private final ResourceMetadata type;
```
Comment (yarpo) [inline am-lib/src/main/java/uk/gov/hmcts/reform/amlib/models/Resource.java]: Maybe `ResourceDefinition` would be a better name?

#minor.

Comment (geek-soft) [reply am-lib/src/main/java/uk/gov/hmcts/reform/amlib/models/Resource.java]: I thought about `ResourceDefinition` as well. Either we change now or later.

File: am-lib/src/main/java/uk/gov/hmcts/reform/amlib/models/ResourceMetadata.java [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package uk.gov.hmcts.reform.amlib.models;
+
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+
+@Data
+@Builder
+@AllArgsConstructor
```
Comment (geek-soft) [inline am-lib/src/main/java/uk/gov/hmcts/reform/amlib/models/ResourceMetadata.java]: Please try hiding constructor. I would like to hide constructors to enforce use of builders. We start here and apply everywhere later.

File: am-lib/src/main/java/uk/gov/hmcts/reform/amlib/repositories/AccessManagementRepository.java [hunk 3]
```diff
@@ -45,4 +45,7 @@
     @SqlQuery("select * from default_permissions_for_roles where service_name = :serviceName and resource_type = :resourceType and resource_name = :resourceName and role_name = :roleName")
     @RegisterConstructorMapper(RoleBasedAccessRecord.class)
     List<RoleBasedAccessRecord> getRolePermissions(String serviceName, String resourceType, String resourceName, String roleName);
+
+    @SqlQuery("select access_management_type from roles where role_name = :roleName")
+    String getRoleAccessType(String roleName);
```
Comment (geek-soft) [inline am-lib/src/main/java/uk/gov/hmcts/reform/amlib/repositories/AccessManagementRepository.java]: Please return enum type. Using both raw and right types brings a mess unnecessarily. Custom mapper should be the right solution. I will require us to use these for other enums as well.

File: am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/FilterResourceIntegrationTest.java [hunk 4]
```diff
@@ -59,11 +70,40 @@ void whenRowExistsAndDoesNotHaveReadPermissionsShouldReturnEnvelopeWithoutData()
...
         String nonExistingResourceId = "lmn";
 
-        FilterResourceResponse result = ams.filterResource(nonExistingUserId, nonExistingResourceId, DATA);
+        FilterResourceResponse result = ams.filterResource(
+            nonExistingUserId, ROLE_NAMES, createResource(nonExistingResourceId));
+
+        assertThat(result).isEqualTo(FilterResourceResponse.builder()
+            .resourceId(resourceId)
+            .data(null)
+            .permissions(ImmutableMap.of(JsonPointer.valueOf(""), CREATE_PERMISSION))
```
Comment (geek-soft) [inline am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/FilterResourceIntegrationTest.java]: How does it work? There is not pre-configuration logic. Roles setup from other tests? Do we truncate roles table before tests?

File: am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/importer/DefaultPermissionIntegrationTest.java [hunk 5]
```diff
@@ -44,7 +46,7 @@ void shouldNotBeAbleToCreateDefaultPermissionWhenRoleDoesNotExist() {
     @Test
     void shouldAddNewEntryIntoDatabaseWhenUniqueEntry() {
         service.addRole(ROLE_NAME, RoleType.RESOURCE, SecurityClassification.PUBLIC, AccessType.ROLE_BASED);
-        service.grantDefaultPermission(createDefaultPermissionGrant(READ_PERMISSION));
+        service.grantDefaultPermission(createDefaultPermissionGrant(JsonPointer.valueOf(ATTRIBUTE), READ_PERMISSION));
```
Comment (geek-soft) [inline am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/importer/DefaultPermissionIntegrationTest.java]: Can we change ATTRIBUTE to be of JsonPointer type? That should be our default type and toStirng is there if string is needed.

Comment (geek-soft) [reply am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/importer/DefaultPermissionIntegrationTest.java]: Rest of changes in that file seems of similar nature, possible candidate for revert.

File: am-lib/src/main/java/uk/gov/hmcts/reform/amlib/AccessManagementService.java [hunk 6]
```diff
@@ -118,28 +119,49 @@ public void revokeResourceAccess(ExplicitAccessMetadata explicitAccessMetadata)
...
-     * @param resourceId   resource ID
-     * @param resourceJson JSON resource
+     * @param userId    accessor ID
+     * @param userRoles user roles
+     * @param resource  envelope {@link Resource} and corresponding metadata
      * @return envelope {@link FilterResourceResponse} with resource ID, filtered JSON and map of permissions if access
      *     to resource is configured, otherwise null.
      */
-    public FilterResourceResponse filterResource(String userId, String resourceId, JsonNode resourceJson) {
+    public FilterResourceResponse filterResource(String userId, Set<String> userRoles, Resource resource) {
```
[the remark this answers is not in the document]
Comment (swalker125) [reply am-lib/src/main/java/uk/gov/hmcts/reform/amlib/AccessManagementService.java]: 🙄

File: am-lib/src/main/java/uk/gov/hmcts/reform/amlib/AccessManagementService.java [hunk 7]
```diff
@@ -118,28 +119,49 @@ public void revokeResourceAccess(ExplicitAccessMetadata explicitAccessMetadata)
...
+            if (jdbi.withExtension(AccessManagementRepository.class,
+                dao -> dao.getRoleAccessType(userRoles.iterator().next())).equals(AccessType.EXPLICIT.name())) {
+                return null;
+            }
+
+            attributePermissions = roleBasedAccess.stream().collect(
+                Collectors.toMap(AttributeAccessDefinition::getAttribute, AttributeAccessDefinition::getPermissions)
+            );
+
+        } else {
```
[the remark this answers is not in the document]
Comment (swalker125) [reply am-lib/src/main/java/uk/gov/hmcts/reform/amlib/AccessManagementService.java]: I think the if else here reduces the total amount of code, and makes everything a bit more common. I only need one return to return either explicit or role based access permissions.

File: am-lib/src/main/java/uk/gov/hmcts/reform/amlib/AccessManagementService.java [hunk 8]
```diff
@@ -118,28 +119,49 @@ public void revokeResourceAccess(ExplicitAccessMetadata explicitAccessMetadata)
...
+                    userRoles.iterator().next()));
+
+            if (roleBasedAccess.isEmpty()) {
+                return null;
+            }
 
-        Map<JsonPointer, Set<Permission>> attributePermissions = explicitAccess.stream().collect(
-            Collectors.toMap(AttributeAccessDefinition::getAttribute, AttributeAccessDefinition::getPermissions)
-        );
+            if (jdbi.withExtension(AccessManagementRepository.class,
```
[the remark this answers is not in the document]
Comment (geek-soft) [reply am-lib/src/main/java/uk/gov/hmcts/reform/amlib/AccessManagementService.java]: Yeah, saw it later on AM-107.

File: am-lib/src/main/java/uk/gov/hmcts/reform/amlib/models/Resource.java [hunk 9]
```diff
@@ -0,0 +1,15 @@
+package uk.gov.hmcts.reform.amlib.models;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+
+@Data
+@Builder
+@AllArgsConstructor
```
Comment (geek-soft) [inline am-lib/src/main/java/uk/gov/hmcts/reform/amlib/models/Resource.java]: Can we remove that constructor? Builder does not need it, I believe it generates private one under the hood.

File: build.gradle [hunk 10]
```diff
@@ -180,6 +180,9 @@ def versions = [
 dependencies {
     compile project(':am-lib')
 
+    annotationProcessor group: 'org.projectlombok', name: 'lombok', version: '1.18.6'
```
Comment (geek-soft) [inline build.gradle]: Why is that needed?

File: src/main/java/uk/gov/hmcts/reform/amapi/models/FilterResource.java [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+import uk.gov.hmcts.reform.amlib.models.Resource;
+
+import java.util.Set;
+
+@Data
+@Builder
+@AllArgsConstructor
```
Comment (geek-soft) [inline src/main/java/uk/gov/hmcts/reform/amapi/models/FilterResource.java]: Is that just needed for the testing service to deserialise request payload?

File: build.gradle (hunk 10 above)
Comment (swalker125) [reply build.gradle]: because we have moved the FilterResourceModel into the am-lib testing service

File: src/main/java/uk/gov/hmcts/reform/amapi/models/FilterResource.java (hunk 11 above)
Comment (swalker125) [reply src/main/java/uk/gov/hmcts/reform/amapi/models/FilterResource.java]: yes

File: build.gradle (hunk 10 above)
Comment (geek-soft) [reply build.gradle]: Good point. I have forgotten about that move. 👍

File: am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/FilterResourceIntegrationTest.java [hunk 12]
```diff
@@ -59,11 +70,36 @@ void whenRowExistsAndDoesNotHaveReadPermissionsShouldReturnEnvelopeWithoutData()
...
+
+        assertThat(result).isEqualTo(FilterResourceResponse.builder()
+            .resourceId(resourceId)
+            .data(DATA)
+            .permissions(ImmutableMap.of(ROOT_ATTRIBUTE, READ_PERMISSION))
+            .build());
+    }
+
+    @Test
+    void whenNoExplicitAccessAndRoleHasExplicitAccessType() {
```
Comment (geek-soft) [inline am-lib/src/test/java/integration/uk/gov/hmcts/reform/amlib/FilterResourceIntegrationTest.java]: Lack of "Should" in test name.
